stdlib/src/collections/deque.mojo

```diff
@@ -0,0 +1,313 @@
+
+        self.tail = len(other)
+
+    fn __moveinit__(inout self, owned existing: Self):
+        """Moves data of an existing deque into a new one.
+
+        Args:
+            existing: The existing deque.
+        """
+        self.data = existing.data
+        self.capacity = existing.capacity
+        self.head = existing.head
+        self.tail = existing.tail
+        self.minlen = existing.minlen
+        self.maxlen = existing.maxlen
+        self.shrink = existing.shrink
+
+    fn __del__(owned self):
+        """Destroys all elements in the deque and free its memory."""
+        for i in range(len(self)):
+            offset = (self.head + i) & (self.capacity - 1)
+            (self.data + offset).destroy_pointee()
+        self.data.free()
+
+    # ===-------------------------------------------------------------------===#
+    # Trait implementations
+    # ===-------------------------------------------------------------------===#
+
+    @always_inline
+    fn __bool__(self) -> Bool:
+        """Checks whether the deque has any elements or not.
+
+        Returns:
+            `False` if the deque is empty, `True` if there is at least one element.
+        """
+        return self.head != self.tail
+
+    @always_inline
+    fn __len__(self) -> Int:
+        """Gets the number of elements in the deque.
+
+        Returns:
+            The number of elements in the deque.
+        """
+        return (self.tail - self.head) & (self.capacity - 1)
+
+    fn __getitem__(ref [_]self, idx: Int) -> ref [self] ElementType:
+        """Gets the deque element at the given index.
+
+        Args:
+            idx: The index of the element.
+
+        Returns:
+            A reference to the element at the given index.
+        """
+        normalized_idx = idx
+
+        debug_assert(
+            -len(self) <= normalized_idx < len(self),
+            "index: ",
+            normalized_idx,
+            " is out of bounds for `Deque` of size: ",
+            len(self),
+        )
+
+        if normalized_idx < 0:
+            normalized_idx += len(self)
+
+        offset = (self.head + normalized_idx) & (self.capacity - 1)
+        return (self.data + offset)[]
+
+    # ===-------------------------------------------------------------------===#
+    # Methods
+    # ===-------------------------------------------------------------------===#
+
+    fn append(inout self, owned value: ElementType):
+        """Appends a value to the right side of the deque.
+
+        Args:
+            value: The value to append.
+        """
+        if len(self) == self.maxlen:
+            (self.data + self.head).destroy_pointee()
+            self.head = (self.head + 1) & (self.capacity - 1)
+
+        (self.data + self.tail).init_pointee_move(value^)
+        self.tail = (self.tail + 1) & (self.capacity - 1)
+
+        if self.head == self.tail:
+            self._realloc(self.capacity << 1)
+
+    fn extend(inout self, owned values: List[ElementType]):
```

Comment:
Hm.. How I understand the current way of Mojo, it only moves the values to the `owned` args on the last use or if explicitly marked by the user with `^`. Eg.:
```mojo
    q = Deque(1, 2)
    l = List(3, 4)

    q.extend(l)
    print(q.__str__()) # Deque(1, 2, 3, 4)
    print(l.__str__())  # [3, 4]

    q.extend(l^)
    print(q.__str__()) # Deque(1, 2, 3, 4, 3, 4)
    # print(l.__str__()) # would not compile, because `l` was moved.

```
Otherwise, it makes an implicit copy. And according to the stdlib style recommendation (https://github.com/modularml/mojo/blob/main/stdlib/docs/style-guide.md#%E2%84%B9%EF%B8%8Fprefer-explicit-copy-constructors-avoid-allowing-implicit-copies) List should not even have the `__copyinit__` constructor to force the users contiously copy the list with `List(other=l)` if they still need it after the `extend()`.

Do I understand it wrong?

Reply:
I did not address this comment in my last commit as I need more information what we would really be aiming to achieve with `inout values`, and do my comments above make sense.